src/pudl_archiver/depositors/depositor.py

```diff
@@ -0,0 +1,336 @@
+        old_datapackage: DataPackage,
+    ) -> tuple[DataPackage, bool]:
+        """Generate new datapackage describing draft deposition in current state."""
+        new_datapackage = self.deposition.generate_datapackage(resources)
+
+        # Add datapackage if it's changed
+        if updated := self._datapackage_worth_changing(
+            old_datapackage, new_datapackage
+        ):
+            datapackage_json = io.BytesIO(
+                bytes(
+                    new_datapackage.model_dump_json(by_alias=True, indent=4),
+                    encoding="utf-8",
+                )
+            )
+            await self.deposition.create_file("datapackage.json", datapackage_json)
+        return new_datapackage, updated
+
+    async def cleanup_after_error(self, e: Exception):
+        """Cleanup draft after an error during an archive run."""
+        await self.deposition.cleanup_after_error(e)
+
+    async def publish(self, run_summary: RunSummary, datapackage_updated: bool):
+        """Cleanup draft after an error during an archive run."""
+        logger.info("Attempting to publish deposition.")
+        if len(run_summary.file_changes) == 0 and not datapackage_updated:
+            logger.info(
+                "No changes detected, kept draft at"
+                f"{self.deposition.get_deposition_link()} for inspection."
+            )
+            return
+        await self.deposition.publish()
+
+    async def get_file(self, filename: str) -> bytes | None:
+        """Get file from deposition.
+
+        Args:
+            filename: Name of file to fetch.
+        """
+        return await self.deposition.get_file(filename)
+
+    async def list_files(self) -> list[str]:
+        """Return list of filenames from previous version of deposition."""
+        return await self.deposition.list_files()
+
+    def get_deposition_link(self) -> Url:
+        """Get URL which points to deposition."""
+        return self.deposition.get_deposition_link()
+
+
+class Depositor(BaseModel):
+    """Wrapper class to manage state of deposition and calling interface."""
+
+    deposition: AbstractDepositorInterface
+    settings: RunSettings
+
+    @classmethod
+    async def get_latest_version(
+        cls,
+        interface: type[AbstractDepositorInterface],
+        dataset: str,
+        session: aiohttp.ClientSession,
+        settings: RunSettings,
+    ) -> "Depositor":
+        """Create a new ZenodoDepositor.
+
+        Args:
+            dataset: Name of dataset to archive.
+            session: HTTP handler - we don't use it directly, it's wrapped in self.request.
+            settings: RunSettings taken from CLI.
+        """
+        interface = await interface.get_latest_version(
+            dataset,
+            session,
+            settings,
+        )
+        return cls(
+            deposition=interface,
+            settings=settings,
+        )
+
+    @asynccontextmanager
+    async def open_draft(self):
```

Comment:
Returns a context manager representing an open draft deposition, so it can cleanly cleanup if there's an error during a run.

Reply:
This is pretty slick!